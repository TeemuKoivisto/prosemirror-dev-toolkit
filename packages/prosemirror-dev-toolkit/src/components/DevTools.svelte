<script lang="ts">
  import { onMount } from 'svelte'
  import type { EditorView } from 'prosemirror-view'
  import type { EditorState, Transaction } from 'prosemirror-state'
  import { setContext } from '$context'

  import FloatingBtn from './FloatingBtn.svelte'
  import FloatingDock from './FloatingDock.svelte'

  import type { ButtonPosition } from '../types'

  export let view: EditorView,
    devToolsExpanded = false,
    buttonPosition: ButtonPosition = 'bottom-right'

  setContext('editor-view', {
    view,
    execCmd(cmd: (state: EditorState, dispatch?: (tr: Transaction) => void) => void) {
      cmd(view.state, view.dispatch)
    },
    replaceEditorContent(state: EditorState) {
      const tr = view.state.tr
      tr.replaceWith(0, view.state.doc.nodeSize - 2, state.doc.content)
      view.dispatch(tr)
    }
  })

  onMount(() => {
    const html = document && document.querySelector('html')
    if (devToolsExpanded && html) {
      html.style.paddingBottom = '341px'
    }
  })

  function handleFloatingBtnClick() {
    devToolsExpanded = true
    const html = document && document.querySelector('html')
    if (html) {
      html.style.paddingBottom = '341px'
    }
  }
  function handleFloatingDockClose() {
    devToolsExpanded = false
    const html = document && document.querySelector('html')
    if (html) {
      html.style.paddingBottom = ''
    }
  }
</script>

<section class="dev-tools">
  {#if devToolsExpanded}
    <FloatingDock onClose={handleFloatingDockClose} />
  {:else}
    <FloatingBtn on:click={handleFloatingBtnClick} {buttonPosition} />
  {/if}
</section>

<style lang="scss">
  * {
    --font-sans: Helvetica Neue, Calibri Light, Roboto, sans-serif;
    --font-family: monaco, Consolas, Lucida Console, monospace;
    --font-small: 11px;
    --font-medium: 13px;
    --font-large: 16px;
    --height-tabs-menu: 48px;
  }
  .dev-tools {
    font-family: var(--font-sans);
    font-size: var(--font-medium);
    font-weight: 400;

    --tree-view-font-family: 'Helvetica Neue', 'Calibri Light', Roboto, sans-serif;
    --tree-view-font-size: 13px;
    --tree-view-left-indent: 0.875em;
    --tree-view-line-height: 1.1;
    --tree-view-key-margin-right: 0.5em;

    --tree-view-base00: #363755;
    --tree-view-base01: #604d49;
    --tree-view-base02: #6d5a55;
    --tree-view-base03: #d1929b;
    --tree-view-base04: #b79f8d;
    --tree-view-base05: #f9f8f2;
    --tree-view-base06: #f7f4f1;
    --tree-view-base07: #faf8f5;
    --tree-view-base08: #fa3e7e;
    --tree-view-base09: #fd993c;
    --tree-view-base0A: #f6bf81;
    --tree-view-base0B: #b8e248;
    --tree-view-base0C: #b4efe4;
    --tree-view-base0D: #85d9ef;
    --tree-view-base0E: #be87ff;
    --tree-view-base0F: #d6724c;

    :global(.svelte-tree-view *) {
      box-sizing: border-box;
    }
  }
</style>
