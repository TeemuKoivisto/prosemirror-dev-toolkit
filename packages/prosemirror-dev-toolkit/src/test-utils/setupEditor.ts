import { DirectEditorProps, EditorView } from 'prosemirror-view'
import { EditorState } from 'prosemirror-state'
import { exampleSetup } from 'prosemirror-example-setup'

import { schema } from './schema'

interface Opts {
  exampleSetup?: boolean
  props?: Omit<DirectEditorProps, 'state'> 
}

const DEFAULT_OPTIONS = {
  exampleSetup: false
}

export function setupEditor(element: HTMLElement, opts: Opts = DEFAULT_OPTIONS) {
  return new EditorView(
    { mount: element },
    {
      state: EditorState.create({
        schema,
        plugins: opts.exampleSetup ? exampleSetup({ schema }) : []
      }),
      ...opts.props
    }
  )
}
