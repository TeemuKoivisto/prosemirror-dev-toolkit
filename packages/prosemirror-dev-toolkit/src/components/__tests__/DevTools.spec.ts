import { render, fireEvent } from '@testing-library/svelte'
import { EditorView } from 'prosemirror-view'
import { vi } from 'vitest'

import DevTools from '../DevTools.svelte'
import { setupEditor } from '$test-utils/setupEditor'

describe('DevTools component', () => {
  let editor: EditorView
  const div = document.createElement('div')
  div.id = 'pm-editor'
  document.body.appendChild(div)
  // polyfillDom()

  beforeAll(() => {
    vi.stubGlobal('prompt', (str: string) => undefined)
  })

  afterEach(() => {
    if (editor) {
      editor.destroy()
    }
    vi.restoreAllMocks()
  })

  it('should render', () => {
    editor = setupEditor(div)
    const results = render(DevTools, {
      props: {
        view: editor,
        buttonPosition: 'top-left',
        devToolsExpanded: true
      }
    })

    expect(results.container).toBeInTheDocument()
    expect(document.body).toMatchSnapshot()
  })

  it('should open by clicking the floating button', async () => {
    editor = setupEditor(div)
    const results = render(DevTools, { props: { view: editor } })
    const button = results.container.querySelector('button')
    expect(button).not.toBeNull()

    await fireEvent.click(button as HTMLElement)
    const snapshotsTabBtn = await results.findByText('SNAPSHOTS')
    await fireEvent.click(snapshotsTabBtn as HTMLElement)

    expect(results.container).toBeInTheDocument()
    expect(document.body).toMatchSnapshot()
  })
})
