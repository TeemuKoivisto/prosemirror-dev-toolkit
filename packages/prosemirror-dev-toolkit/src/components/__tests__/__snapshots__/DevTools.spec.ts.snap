// Vitest Snapshot v1

exports[`DevTools.svelte > should open by clicking the floating button 1`] = `
<body>
  <div
    class="ProseMirror"
    contenteditable="true"
    id="pm-editor"
    translate="no"
  >
    <p>
      <br
        class="ProseMirror-trailingBreak"
      />
    </p>
  </div>
  <div>
    <section
      class="dev-tools s-DevTools-3laa75"
    >
      <div
        class="floating-dock-wrapper s-FloatingDock-14aalj4"
      >
        <div
          class="paste-modal s-PasteModal-1516w55 hidden"
        >
          <div
            class="modal-bg s-PasteModal-1516w55"
          />
           
          <form
            class="paste-content s-PasteModal-1516w55"
          >
            <fieldset
              class="s-PasteModal-1516w55"
            >
              <div
                class="submit-container s-PasteModal-1516w55"
              >
                <button
                  class="s-PasteModal-1516w55"
                >
                  Submit
                </button>
              </div>
               
              <legend
                class="s-PasteModal-1516w55"
              >
                Doc
              </legend>
               
              <textarea
                class="s-PasteModal-1516w55"
              />
            </fieldset>
          </form>
        </div>
        <!--&lt;PasteModal&gt;-->
         
        <div
          class="floating-dock s-FloatingDock-14aalj4"
          style="top: 50%; height: 50%;"
        >
          <div
            class="resizing-div s-FloatingDock-14aalj4"
          />
           
          <div
            class="container s-FloatingDock-14aalj4"
          >
            <div
              class="s-FloatingDock-14aalj4"
            >
              <button
                class="copy-btn s-FloatingDock-14aalj4"
              >
                Copy
              </button>
               
              <button
                class="save-btn s-FloatingDock-14aalj4"
              >
                Save
              </button>
               
              <button
                class="import-btn s-FloatingDock-14aalj4"
              >
                Import
              </button>
               
              <button
                class="paste-btn s-FloatingDock-14aalj4"
              >
                Paste
              </button>
               
              <button
                aria-label="Close dev-toolkit"
                class="close-btn s-FloatingDock-14aalj4"
              >
                X
              </button>
            </div>
             
            <input
              accept=".json"
              class="s-FloatingDock-14aalj4"
              multiple=""
              style="display: none;"
              type="file"
            />
             
            <ul
              class="tabs-menu s-TabsMenu-1on5hou"
            >
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  STATE
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  HISTORY
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  PLUGINS
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  SCHEMA
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  STRUCTURE
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou active"
                >
                  SNAPSHOTS
                </button>
              </li>
            </ul>
            <!--&lt;TabsMenu&gt;-->
             
            <section
              class="split-view s-SplitView-1ar0nx3"
            >
               
              <div
                class="right-panel s-SnapshotsTab-1cy4q67"
                slot="right"
              >
                <div
                  class="no-snapshots s-SnapshotsTab-1cy4q67"
                >
                  Save snapshots by clicking "Save" button.
                </div>
              </div>
            </section>
            <!--&lt;SplitView&gt;-->
            <!--&lt;SnapshotsTab&gt;-->
          </div>
        </div>
      </div>
      <!--&lt;FloatingDock&gt;-->
    </section>
    <!--&lt;DevTools&gt;-->
  </div>
</body>
`;

exports[`DevTools.svelte > should render 1`] = `
<body>
  <div
    class="ProseMirror"
    contenteditable="true"
    id="pm-editor"
    translate="no"
  >
    <p>
      <br
        class="ProseMirror-trailingBreak"
      />
    </p>
  </div>
  <div>
    <section
      class="dev-tools s-DevTools-3laa75"
    >
      <div
        class="floating-dock-wrapper s-FloatingDock-14aalj4"
      >
        <div
          class="paste-modal s-PasteModal-1516w55 hidden"
        >
          <div
            class="modal-bg s-PasteModal-1516w55"
          />
           
          <form
            class="paste-content s-PasteModal-1516w55"
          >
            <fieldset
              class="s-PasteModal-1516w55"
            >
              <div
                class="submit-container s-PasteModal-1516w55"
              >
                <button
                  class="s-PasteModal-1516w55"
                >
                  Submit
                </button>
              </div>
               
              <legend
                class="s-PasteModal-1516w55"
              >
                Doc
              </legend>
               
              <textarea
                class="s-PasteModal-1516w55"
              />
            </fieldset>
          </form>
        </div>
        <!--&lt;PasteModal&gt;-->
         
        <div
          class="floating-dock s-FloatingDock-14aalj4"
          style="top: 50%; height: 50%;"
        >
          <div
            class="resizing-div s-FloatingDock-14aalj4"
          />
           
          <div
            class="container s-FloatingDock-14aalj4"
          >
            <div
              class="s-FloatingDock-14aalj4"
            >
              <button
                class="copy-btn s-FloatingDock-14aalj4"
              >
                Copy
              </button>
               
              <button
                class="save-btn s-FloatingDock-14aalj4"
              >
                Save
              </button>
               
              <button
                class="import-btn s-FloatingDock-14aalj4"
              >
                Import
              </button>
               
              <button
                class="paste-btn s-FloatingDock-14aalj4"
              >
                Paste
              </button>
               
              <button
                aria-label="Close dev-toolkit"
                class="close-btn s-FloatingDock-14aalj4"
              >
                X
              </button>
            </div>
             
            <input
              accept=".json"
              class="s-FloatingDock-14aalj4"
              multiple=""
              style="display: none;"
              type="file"
            />
             
            <ul
              class="tabs-menu s-TabsMenu-1on5hou"
            >
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou active"
                >
                  STATE
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  HISTORY
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  PLUGINS
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  SCHEMA
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  STRUCTURE
                </button>
              </li>
               
              <li
                class="s-TabsMenu-1on5hou"
              >
                <button
                  class="s-TabsMenu-1on5hou"
                >
                  SNAPSHOTS
                </button>
              </li>
            </ul>
            <!--&lt;TabsMenu&gt;-->
             
            <section
              class="split-view s-SplitView-1ar0nx3"
            >
              <div
                class="left-panel s-StateTab-7b9rm4"
                slot="left"
              >
                <div
                  class="top-row s-StateTab-7b9rm4"
                >
                  <h2
                    class="s-StateTab-7b9rm4"
                  >
                    Current doc
                  </h2>
                   
                  <button
                    class="s-Button-ss8iwa"
                  >
                    log
                  </button>
                  <!--&lt;Button&gt;-->
                </div>
                 
                <ul
                  class="tree-view svelte-tree-view s-TreeView-6v753y"
                >
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[0]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      type
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="string"
                      role="presentation"
                    >
                      "doc"
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                      <button
                        class="log-copy-button s-TreeViewNode-wbyn6n"
                      >
                        log
                      </button>
                       
                      <button
                        class="log-copy-button s-TreeViewNode-wbyn6n"
                      >
                        copy
                      </button>
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n collapsed"
                    data-tree-id="[1]"
                  >
                    <button
                      class="arrow-btn collapsed s-TreeViewNode-wbyn6n"
                    >
                      ▶
                    </button>
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n has-children"
                      role="presentation"
                    >
                      content
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n has-children"
                      data-type="array"
                      role="presentation"
                    >
                       [] 1 items
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                      <button
                        class="log-copy-button s-TreeViewNode-wbyn6n"
                      >
                        log
                      </button>
                       
                      <button
                        class="log-copy-button s-TreeViewNode-wbyn6n"
                      >
                        copy
                      </button>
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                </ul>
                <!--&lt;TreeView&gt;-->
              </div>
               
              <div
                class="right-panel s-StateTab-7b9rm4"
                slot="right"
              >
                <div
                  class="top-row s-StateTab-7b9rm4"
                >
                  <h2
                    class="s-StateTab-7b9rm4"
                  >
                    Selection
                  </h2>
                   
                  <button
                    class="selection-btn s-Button-ss8iwa"
                  >
                    <span
                      class="caret-icon s-StateTab-7b9rm4"
                    />
                  </button>
                  <!--&lt;Button&gt;-->
                </div>
                 
                <ul
                  class="tree-view svelte-tree-view s-TreeView-6v753y"
                >
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[0]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      jsonID
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="string"
                      role="presentation"
                    >
                      "text"
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[1]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      empty
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="boolean"
                      role="presentation"
                    >
                      true
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[2]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      anchor
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="number"
                      role="presentation"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[3]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      from
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="number"
                      role="presentation"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[4]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      head
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="number"
                      role="presentation"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-TreeViewNode-wbyn6n"
                    data-tree-id="[5]"
                  >
                     
                    <div
                      class="node-key s-TreeViewNode-wbyn6n p-left"
                      role="presentation"
                    >
                      to
                      :
                    </div>
                     
                    <div
                      class="node-value s-TreeViewNode-wbyn6n"
                      data-type="number"
                      role="presentation"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-TreeViewNode-wbyn6n"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                </ul>
                <!--&lt;TreeView&gt;-->
                 
                <div
                  class="s-StateTab-7b9rm4"
                >
                  <h2
                    class="s-StateTab-7b9rm4"
                  >
                    Active marks
                  </h2>
                   
                  <div
                    class="no-marks s-StateTab-7b9rm4"
                  >
                    No active marks
                  </div>
                </div>
                 
                <div
                  class="s-StateTab-7b9rm4"
                >
                  <h2
                    class="s-StateTab-7b9rm4"
                  >
                    Document stats
                  </h2>
                   
                  <ul
                    class="tree-view svelte-tree-view s-TreeView-6v753y"
                  >
                    <li
                      class="row s-TreeViewNode-wbyn6n"
                      data-tree-id="[0]"
                    >
                       
                      <div
                        class="node-key s-TreeViewNode-wbyn6n p-left"
                        role="presentation"
                      >
                        nodeSize
                        :
                      </div>
                       
                      <div
                        class="node-value s-TreeViewNode-wbyn6n"
                        data-type="number"
                        role="presentation"
                      >
                        4
                      </div>
                       
                      <div
                        class="buttons s-TreeViewNode-wbyn6n"
                      >
                         
                      </div>
                    </li>
                     
                    
                    <!--&lt;TreeViewNode&gt;-->
                    <li
                      class="row s-TreeViewNode-wbyn6n"
                      data-tree-id="[1]"
                    >
                       
                      <div
                        class="node-key s-TreeViewNode-wbyn6n p-left"
                        role="presentation"
                      >
                        childCount
                        :
                      </div>
                       
                      <div
                        class="node-value s-TreeViewNode-wbyn6n"
                        data-type="number"
                        role="presentation"
                      >
                        1
                      </div>
                       
                      <div
                        class="buttons s-TreeViewNode-wbyn6n"
                      >
                         
                      </div>
                    </li>
                     
                    
                    <!--&lt;TreeViewNode&gt;-->
                  </ul>
                  <!--&lt;TreeView&gt;-->
                </div>
              </div>
            </section>
            <!--&lt;SplitView&gt;-->
            <!--&lt;StateTab&gt;-->
          </div>
        </div>
      </div>
      <!--&lt;FloatingDock&gt;-->
    </section>
    <!--&lt;DevTools&gt;-->
  </div>
</body>
`;
